### ISSUE TITLE
Feature request: easier access to tensor de-allocation information

### ISSUE THREAD

**Issue by yaroslavvb** [tensorflow/tensorflow#6716]
TLDR; to debug TensorFlow out-of-memory situations one needs to see tensor de-allocation info.

You can see allocation stats in timeline, but without de-allocation info you can't calculate peak memory. Currently getting peak memory is possible by:

1. Hacking TensorFlow to print deallocation messages with timestamps as [here](https://github.com/yaroslavvb/tensorflow/commit/5d4cd97c0a73e91ee37c025cd7a62fb46aae76a0)

2. A bunch of regular expression to parse `__LOG_MEMORY__` messages as in [here](https://github.com/yaroslavvb/notebooks/blob/master/saving%20memory%20by%20using%20functions.ipynb)

Since this needs building your own version of tensorflow, this is not accessible to most people. Perhaps this can be remedied by adding deallocation events to session run timeline? @michaelisard 

Some recent places this issue came up:
http://stackoverflow.com/questions/41517145/outer-product-based-conv-filters-consume-disproportionately-high-memory
http://stackoverflow.com/questions/41496251/fitting-large-matrix-calculations-into-memory-when-using-tensorflow
http://stackoverflow.com/questions/41451273/tensorflow-specifying-storage-of-layer-activations
http://stackoverflow.com/questions/40190510/tensorflow-how-to-log-gpu-memory-vram-utilization/40197094#comment70145571_40197094
https://github.com/tensorflow/tensorflow/issues/6019#issuecomment-268037463

**Comment by michaelisard** [comment]
I will take a look but it will take me a few days to get to it. The reason it isn't trivial to put it in the timelines is that deallocation is done when the refcount falls to zero, not by an explicit call at a site that has a reference to the timeline datastructure.

**Comment by yaroslavvb** [comment]
This would also help with benchmarks which last week [were modified](https://github.com/tensorflow/tensorflow/commit/99d4517a) to store maximum memory usage . As far as I can tell, the memory parsing from timeline just adds up all allocations, so doesn't actually represent peak memory @ebrevdo

**Comment by michaelisard** [comment]
The [timeline](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/client/timeline.py) code does try to figure out the deallocations and compute peak usage, based on essentially reconstituting the graph from the logs. It would be better to do it properly, but I don't think it's as bad as you fear. @prb12 can correct me if I'm wrong.

**Comment by prb12** [comment]
As @michaelisard  says, the Timeline class _does_ have some code in to try and track the refcounts of Tensors using information in the trace, but it is very unreliable since the only information it has is is the string names of inputs to each operation (from the ASCII timeline label).  These aren't robust to graph partitioning, rewriting and constant folding etc., and this is why the memory histograms are disabled by default.

Now that it is possible to retrieve the rewritten graphdefs, tf.Timeline could probably do a much better job of reference count tracking  - but as Michael says, it would be better to actually track the deallocations.  Unfortunately these can happen asynchronously at arbitrary times, so the protobuf format would need extending.   There is also the subtlety of GPU allocations and deallocations happening long before the CUDA kernels actually execute, which potentially makes the UI a little confusing. (not an issue for simply recording peak memory usage)

**Comment by yaroslavvb** [comment]
This may also help solving the mysterious issue of TF running out of memory on step xyz, even though there are no changes to graph between steps. I've ran across it, and also [here](http://stackoverflow.com/questions/41540073/why-the-gpu-memory-is-not-the-same)

**Comment by prb12** [comment]
There are numerous sources of non-determinism in memory allocation (and op dispatch), but more importantly, many people forget that there are often multiple steps running at the same time (e.g. if your model involves queues), does checkpointing, etc.

**Comment by yaroslavvb** [comment]
It seems hard to get accurate deallocation information even from VLOG messages.

The problem is that you can have multiple allocations with the same allocation id, but different step ids in the same `.run` call. The deallocation message does not include step_id, so it's ambiguous.

Here's an example "leak" I spent this morning debugging

```
2017-01-26 10:13:30: I tensorflow/core/framework/log_memory.cc:35] __LOG_MEMORY__ MemoryLogTensorAllocation { step_id: -6 kernel_name: "Unknown (from Proto)" tensor { dtype: DT_FLOAT shape { } allocation_description { requested_bytes: 4 allocated_bytes: 4 allocator_name: "cpu" allocation_id: 6 has_single_reference: true ptr: 5045586496 } } }

2017-01-26 10:13:30: I tensorflow/core/framework/log_memory.cc:35] __LOG_MEMORY__ MemoryLogTensorDeallocation { allocation_id: 6 allocator_name: "cpu" }

2017-01-26 10:13:30: I tensorflow/core/framework/log_memory.cc:35] __LOG_MEMORY__ MemoryLogTensorAllocation { step_id: 2 kernel_name: "a1" tensor { dtype: DT_FLOAT shape { dim { size: 250000 } } allocation_description { requested_bytes: 1000000 allocated_bytes: 1000192 allocator_name: "gpu_bfc" allocation_id: 6 has_single_reference: true ptr: 30078652160 } } }

2017-01-26 10:13:30: I tensorflow/core/framework/log_memory.cc:35] __LOG_MEMORY__ MemoryLogTensorDeallocation { allocation_id: 6 allocator_name: "gpu_bfc" }

```

**Comment by yaroslavvb** [comment]
I added some heuristics to deal with cases like above and packaged it up as standalone script at https://github.com/yaroslavvb/memory_util

**Comment by michaelisard** [comment]
I have given this a little thought but not yet looked into the details of the implementation. It looks as if it should be ok to add the step_id to the tracked metadata, so that you won't have to use heuristics to guess it.

As for putting deallocation into the timeline, I have one idea which is a bit half-baked and has not been subjected to code review. Before I type it in, I'd like feedback on whether it would solve your problem or not.

I could add a thread-local callback to the tracking allocator, such that deallocations trigger the callback. Then, when memory tracking is enabled, each Op would install a callback before it began executing, and remove it after it completed. Any tensor that was freed on that thread while the Op was executing would be added to the timeline of the Op. This would obviously fail to work if deallocations happened on an intra-op threadpool, but hopefully that is rare or nonexistent.

What do you think?

**Comment by yaroslavvb** [comment]
So the problem I'm trying to solve is that we have models that use too much memory, and in order to have confidence that our optimizations are working, we want to be able to track down "MemoryExceeded" error to a particular sequence of allocations and deallocations.

For your proposal, my question would be how robust it is. What do you mean mean by "hopefully it's rare"? If it's pretty rare, then I think it would simplify/robustify things if I could build the memory usage timeline using step-metadata instead of parsing stderr.

BTW, another thing I came across with LOG_MEMORY messages is that sometimes I get deallocation messages even when there's no mention of this allocation_id in this run call. Can things be de-allocated in a run call if they weren't allocated in the same session.run call?

IE, message like this

`     41:2017-01-26 15:06:29: I tensorflow/core/framework/log_memory.cc:35] __LOG_MEMORY__ MemoryLogTensorDeallocation { allocation_id: 29330 allocator_name: "cuda_host_bfc" }
`

**Comment by michaelisard** [comment]
The issue would arise if temporaries got allocated by threads on the Eigen threadpool. So it shouldn't happen for GPU allocations. For CPU I haven't looked at all the ops: someone could do it, but it seems unlikely to be performance-friendly.

Anything held in the resource manager may persist across run calls, and therefore be allocated and deallocated by different calls. E.g., a tensor in a queue, or a variable that is being replaced by one of a different shape. Off-hand I don't think other allocations should cross run calls, but I am not 100% sure.

**Comment by yaroslavvb** [comment]
BTW, here's an example of things we are using it for, looking at timelines like attached (this is memory usage timeline for regular tf.gradients vs. sublinear memory gradients from https://arxiv.org/abs/1604.06174 for 100 node chain)

<img width="413" alt="default" src="https://cloud.githubusercontent.com/assets/23068/22355705/1f1ddd88-e3e1-11e6-9888-5a5d5214af3c.png">
<img width="413" alt="screenshot 2017-01-26 16 00 56" src="https://cloud.githubusercontent.com/assets/23068/22355707/22164b74-e3e1-11e6-9cf6-148de2ebe981.png">

**Comment by yaroslavvb** [comment]
I was seeing those "orphan" deallocations in https://github.com/tensorflow/models/tree/master/resnet which uses queues...although `cuda_host_bfc` seems to rule that out, queues should be using CPU 

To summarize, I think any change that would make memory timelines easier to build would be welcome. My current vlog memory parsing logic seems to mostly work, but it breaks when vlog output gets too large (ie, vlog=2 doesn't work because of gazillion GPU polling messages), or when protobuf text output changes

**Comment by michaelisard** [comment]
For the orphan allocation, I think the most likely is that someone is calling AllocateRaw and forgetting to call LogMemory::RecordRawAllocation at the same time. The easiest way to debug it is probably to add a log statement in TrackingAllocator::AllocateRaw. If there's an orphan deallocation of size Q, grep through for all the allocations of size Q and look for the one with no matching __LOG_MEMORY__. Then report the bug :)

**Comment by yaroslavvb** [comment]
Hi @yuefengz , I see from commit messages you've been doing memory work, do you know what's the status of memory tracking here?

In particular I'm wondering the best way to get timeline of memory allocation. We've needed this functionality to use memory-efficient-gradients package that we are planning to release.

I've been maintaining [memory_util](https://github.com/yaroslavvb/memory_util) which gives a timeline of Tensor Allocations from log messages from [log_memory.cc](https://github.com/tensorflow/tensorflow/blob/754048a0453a04a761e112ae5d99c149eb9910dd/tensorflow/core/framework/log_memory.cc#L35) , but TF 1.0.1 -> TF 1.1 update seems to have broken it. Now this module prints deallocation messages with missing allocation id.

**Comment by yuefengz** [comment]
Hi @yaroslavvb , the memory tracking that I've implemented is mostly for recording the memory usages, include temp memory, output memory and persistent memory. This is useful for building a cost model but it does't account for tracking memory allocation and deallocation.

**Comment by yaroslavvb** [comment]
relevant commit: github.com/tensorflow/tensorflow/commit/86238e8d

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by alextp** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
